### PR TITLE
Follow-up to PR #20: remove hardcoded paths + training improvements

### PR DESCRIPTION
**Pull request by sonalivyascse19-stack**

### CODE REVIEW

**Comment by sonalivyascse19-stack** [comment]
1) Made training deterministic + reproducible

- Added a --seed CLI arg and a set_seed() helper.
- Set seeds for random, numpy, torch, and torch.cuda.
- Set torch.backends.cudnn.deterministic=True and benchmark=False to reduce run-to-run variation.

Result: repeated runs with --seed 42 are now consistent (loss + metrics match closely).

2) Fixed output scaling so the model doesn’t “ignore” small-magnitude targets (especially q_spigot)

- Confirmed that q_spigot values were small and could be underweighted if we train naïvely with standard MSE across multiple outputs.
- In ModelController.train_model() we compute dataset-level mean/std of targets (for the last timestep, across all buckets) and train using normalized targets:
- targets_n = (targets_last - y_mean) / y_std
- output_n = (output - y_mean) / y_std

This makes each output contribute more equally to loss (variance-normalized), which improved q_spigot NSE notably compared to earlier runs.

3) Tuned data generation / config to make q_spigot less “always on” and more learnable

Investigated H_spigot/H_bucket and initial water levels; originally many buckets started above the spigot and produced nearly-always-nonzero q_spigot.
Adjusted rH_spigot range in the config so fewer buckets always spill via spigot, increasing variability/structure in q_spigot.

Verified by printing:

- ratio stats (H_spigot/H_bucket)
- % initial water above spigot
- % nonzero q_spigot

This made the learning problem more realistic and helped improve spigot-related performance.

4) Added optional “mass constraint” hook (tested but kept conservative)

Added YAML support:

loss:
  lambda_mass: <value>

Experimented with larger values (e.g., 3.0) and saw it could hurt NSE, so backed off (e.g., 0.01 / 0.0) and kept it as an optional knob rather than forcing it.
Quality-of-life / robustness changes (not performance directly, but helped run/testing)

5) Removed hardcoded execution assumptions in the runner

Cleaned up run/run_deep_bucket_lab.py to use argparse properly and avoid hardcoded config paths beyond a sensible default.

Made it easy to run:

python run/run_deep_bucket_lab.py --config configuration/tune_seq48_h64.yml --seed 42

6) Fixed the UH CSV path issue so runs work from any working directory

- The run initially crashed with an OSError because unit_hydrograph_distribution_file was treated relative to wherever we launched Python.
- Updated src/data_generation.py to resolve the CSV path relative to repo root when a relative path is provided (and raise a clear FileNotFoundError if missing).
- This prevents “works on my machine” path failures.


Net effect  after changes:

- q_total NSE improved into clearly positive territory (~0.17–0.18).
- q_overflow NSE improved (~0.12–0.13).
- q_spigot NSE improved from strongly negative to near-zero / small positive in many runs (mean ~0.00–0.01, depending on config).

File: configuration/tune_seq48_h64.yml [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+      - 0.16
+      ET_parameter:
+      - 2.2447
+      - 9.9807e-5
+      - 0.0016
+      H_bucket:
+      - 0.19
+      - 0.21
+      rH_spigot:
+      - 0.60
```
Comment (quinnylee) [inline configuration/tune_seq48_h64.yml:148]: Would love to see any graphs if you have any of how `q_spigot` and `q_overflow` change over time with these new `rH_spigot` values

File: src/model_controller.py [hunk 2]
```diff
@@ -4,95 +4,178 @@
...
+                    data_loss = criterion(output_n, targets_n)
+
+                    # Optional mass penalty in PHYSICAL space
+                    if use_mass_loss and lambda_mass > 0.0:
+                        # Inverse-transform selected outputs to physical units
+                        q_sp = output[:, idx_spigot] * out_scales[idx_spigot] + out_means[idx_spigot]
+                        q_ov = output[:, idx_overflow] * out_scales[idx_overflow] + out_means[idx_overflow]
+                        q_to = output[:, idx_total] * out_scales[idx_total] + out_means[idx_total]
+
+                        mass_err = q_to - (q_sp + q_ov)
```
Comment (quinnylee) [inline src/model_controller.py]: I don't think this mass balance section works as intended. Looking through my code (although I don't exactly remember my rationale for everything), I think `q_overflow` and `q_spigot` are coming out in units of length per time step (meters/hour I think? It's coming out in _yield_, not discharge. That's the important part). I think I might have done this to simplify the bucket simulation.

It is true that at first, `q_total = q_overflow + q_spigot`.
However, `q_total` undergoes two transformations: 
1.  A unit transformation to transform yield into discharge (length^3 / time, m^3/hr)
2.  A unit hydrograph transformation. Basically, total outflow gets time lagged a little bit, since streamflow does not peak instantaneously after the initial rainfall event. To conceptualize this physically, maybe think about how when it starts raining, it takes a little while before all the water shows up in the gutter on the side of the street. (Jonathan will have a better and more correct explanation for you if you are curious)

So calculating mass error from as `q_total - (q_spigot + q_overflow)` is not correct. 

Mass balance is a fun idea though, and something I had trouble implementing successfully, even in the physical simulation. With a time step of an hour, the physical processes and their differential equations that expect continuous behavior kind of stop working well.

However, it would be fun to look at from an ML model performance perspective. You will probably need to implement some extra output variable that represents water that has rained into the bucket but hasn't flowed out for some reason (like if the spigot is too high or something). If it doesn't cause any data leakage issues to go back and look at your input values, you could take the `precip` variable (I don't remember what unit it's in, probably length in meters) and subtract (`q_spigot` + `q_overflow` + `the secret extra output variable that you will invent`) from it to get an error in units of length.

Comment (quinnylee) [reply src/model_controller.py]: Also just remembered that there is water lost through ET and infiltration. Adding ET into the mass balance equation is easy because it's an input variable. Infiltration is a bit trickier since it's an internal physical simulation variable, but loss through infiltration is (as I remember) very small

File: src/model_controller.py [hunk 3]
```diff
@@ -4,95 +4,178 @@
...
-                    loss = criterion(output, targets)
+
+                    data = data.to(self.device)
+                    targets = targets.to(self.device)
+
+                    output = self.lstm(data)          # (batch, out_dim) in SCALED space
+                    targets_last = targets[:, -1, :]  # (batch, out_dim) in SCALED space
+
+                    # Normalize across outputs using dataset-level stats (SCALED space)
+                    targets_n = (targets_last - y_mean) / y_std
```
Comment (quinnylee) [inline src/model_controller.py]: This is me not understanding ML:

I believe the inputs were already normalized before sending them into the LSTM. Why do the outputs need to be normalized too?

Comment (sonalivyascse19-stack) [reply src/model_controller.py]: Good point! Since StandardScaler is already applied in make_data_loader(), the additional normalization was double-normalizing the data. I've removed this redundant step - the model now uses the scaled outputs directly for MSE loss. Training still works well (loss: 0.997→0.817 over 8 epochs).

File: src/model_controller.py (hunk 2 above)
Comment (sonalivyascse19-stack) [reply src/model_controller.py]: Thanks for catching this! You're right - q_total undergoes unit hydrograph transformation (time-lag convolution) so it's not simply equal to q_spigot + q_overflow. I've removed the mass balance penalty entirely in commit aae6430 since the constraint as implemented was physically incorrect.

File: configuration/tune_seq48_h64.yml (hunk 1 above)
Comment (sonalivyascse19-stack) [reply configuration/tune_seq48_h64.yml:148]: precip_vs_outflows
q_spigot_overflow_timeseries
